minard/RSTools.py

```diff
@@ -0,0 +1,528 @@
+def get_RS_reports(criteria=None, run_min=None, run_max=None, limit=None):
+    '''Get run-selection tables in a run range. If duplicate tables, only keeps one
+    (takes one with latest version, and if they have the same version, the one with
+    the latest timestamp).'''
+
+    # Get tables in descending order, with hardcoded max at 100 to avoid slow down if misused
+    query = "SELECT meta_data, name, timestamp FROM run_selection WHERE type = 'RS_REPORT'".format(criteria)
+    conditions = []
+    if criteria is not None:
+        conditions.append("criteria = '{}'".format(str(criteria)))
+    if run_min is not None:
+        conditions.append("run_min >= {}".format(int(run_min)))
+    if run_max is not None:
+        conditions.append("run_max <= {}".format(int(run_max)))
+
+    if len(conditions) > 0:
+        for i in range(0, len(conditions)):
+            query += " AND " + conditions[i]
+
+    query += " ORDER BY run_min DESC"
+    if limit is not None:
+        query += " LIMIT {}".format(limit)
+
+    conn = engine_nl.connect()
+    resultQuery = conn.execute(query)
+
+    rs_tables_list = []
+    for row in resultQuery.fetchall():
+        tempt_dict = {}
+        tempt_dict['meta_data'] = row[0]
+        tempt_dict['name'] = row[1]
+        tempt_dict['timestamp'] = row[2]
+        rs_tables_list.append(tempt_dict)
+
+    if len(rs_tables_list) == 0:
+        print("ERROR: No RS table found for criteria tag '{}', run min {} and run_max {}, ".format(criteria, run_min, run_max))
+        return False
+    
+    # Repackage and check for duplicates
+    rs_tables = {}
+    for table in rs_tables_list:
+        run_number = table['meta_data']['run_range'][0]
+        criteria = table['meta_data']['index']
+        if run_number not in rs_tables:
+            rs_tables[run_number] = {}
+            rs_tables[run_number][criteria] = table
+        else:
+            if criteria not in rs_tables[run_number]:
+                rs_tables[run_number][criteria] = table
+            else:
+                # There is a duplicate
+                if decide_replace_table(rs_tables[run_number][criteria], table):
+                    rs_tables[run_number][criteria] = table
+
+    return rs_tables
+
+def get_filtered_RS_tables(run_max, offset, limit, result, criteria):
+    '''Download run-selection tables in range, and keep only ones with desired result'''
+
+    # Download a few more runs than limit, in case some don't meet conditions
+    rs_tables = get_RS_reports(criteria=criteria, run_max=run_max, limit = (offset + limit) * 2, table_type='RS_REPORT')
+    if rs_tables is False:
+        return False, True
+
+    no_more_tables = False
+    if len(rs_tables) < ((offset + limit) * 2):
+        no_more_tables = True
+
+    ### Check conditions ###
+
+    # In RS tables, the final result is written as True if it passed,
+    # False if it failed, and None if it was purgatoried.
+    resultMapping = {'Pass': True, 'Purgatory': None, 'Fail': False}
```

Comment:
Does this need an "All" key?

Reply:
No, I used an if condition to deal with the "All" case, since that basically means no filter on the result.